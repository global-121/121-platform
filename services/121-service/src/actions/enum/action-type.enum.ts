export enum ActionType {
  startBlockNewPayment = 'start-block-new-payment',
  endBlockNewPayment = 'end-block-new-payment',
}
