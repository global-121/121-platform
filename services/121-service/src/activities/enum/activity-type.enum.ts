export enum ActivityTypeEnum {
  DataChange = 'data-change',
  FinancialServiceProviderChange = 'financial-service-provider-change',
  Message = 'message',
  Note = 'note',
  StatusChange = 'status-change',
  Transaction = 'transaction',
}
