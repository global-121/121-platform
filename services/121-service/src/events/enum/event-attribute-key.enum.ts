export enum EventAttributeKeyEnum {
  reason = 'reason',
  oldValue = 'oldValue',
  newValue = 'newValue',
  fieldName = 'fieldName',
  duplicateWithRegistrationId = 'duplicateWithRegistrationId',
  duplicateWithRegistrationProgramId = 'duplicateWithRegistrationProgramId',
}
