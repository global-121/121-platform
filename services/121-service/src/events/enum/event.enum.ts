export enum EventEnum {
  registrationDataChange = 'registrationDataChange',
  financialServiceProviderChange = 'financialServiceProviderChange',
  registrationStatusChange = 'registrationStatusChange',
  ignoredDuplicate = 'ignoredDuplicate',
}
