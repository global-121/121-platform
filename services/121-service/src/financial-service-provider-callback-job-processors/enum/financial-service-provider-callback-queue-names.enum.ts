export enum FinancialServiceProviderCallbackQueueNames {
  safaricomTransferCallback = 'safaricomTransferCallback',
  safaricomTimeoutCallback = 'safaricomTimeoutCallback',
}
