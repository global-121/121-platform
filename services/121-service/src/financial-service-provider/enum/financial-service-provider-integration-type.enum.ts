export enum FinancialServiceProviderIntegrationType {
  api = 'api',
  csv = 'csv',
  xml = 'xml',
}
