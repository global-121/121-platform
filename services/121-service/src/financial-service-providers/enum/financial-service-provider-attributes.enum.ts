export enum FinancialServiceProviderAttributes {
  phoneNumber = 'phoneNumber',
  nationalId = 'nationalId',
  fullName = 'fullName',
  addressStreet = 'addressStreet',
  addressHouseNumber = 'addressHouseNumber',
  addressHouseNumberAddition = 'addressHouseNumberAddition',
  addressPostalCode = 'addressPostalCode',
  addressCity = 'addressCity',
  bankAccountNumber = 'bankAccountNumber',
  whatsappPhoneNumber = 'whatsappPhoneNumber',
}
