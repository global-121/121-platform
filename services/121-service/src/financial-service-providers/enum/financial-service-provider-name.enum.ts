export enum FinancialServiceProviderName {
  intersolveVoucherWhatsapp = 'Intersolve-voucher-whatsapp',
  intersolveVoucherPaper = 'Intersolve-voucher-paper',
  intersolveVisa = 'Intersolve-visa',
  intersolveJumboPhysical = 'Intersolve-jumbo-physical',
  africasTalking = 'Africas-talking',
  belcash = 'BelCash',
  vodacash = 'VodaCash',
  bobFinance = 'BoB-finance',
  ukrPoshta = 'UkrPoshta',
  safaricom = 'Safaricom',
  commercialBankEthiopia = 'Commercial-bank-ethiopia',
  excel = 'Excel',
}

export enum FinancialServiceProviderConfigurationEnum {
  password = 'password',
  username = 'username',
  columnsToExport = 'columnsToExport',
  columnToMatch = 'columnToMatch',
  brandCode = 'brandCode',
  displayName = 'displayName',
  coverLetterCode = 'coverLetterCode',
}

export const FinancialServiceProviderConfigurationMapping: {
  [key in FinancialServiceProviderName]?: any;
} = {
  [FinancialServiceProviderName.intersolveVoucherWhatsapp]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVoucherPaper]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVisa]: [
    FinancialServiceProviderConfigurationEnum.brandCode,
    FinancialServiceProviderConfigurationEnum.coverLetterCode,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveJumboPhysical]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.africasTalking]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.belcash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.vodacash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.bobFinance]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.ukrPoshta]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.safaricom]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.commercialBankEthiopia]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.excel]: [
    FinancialServiceProviderConfigurationEnum.columnsToExport,
    FinancialServiceProviderConfigurationEnum.columnToMatch,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
};
