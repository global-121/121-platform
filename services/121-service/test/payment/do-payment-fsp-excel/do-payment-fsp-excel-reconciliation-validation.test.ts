import { HttpStatus } from '@nestjs/common/enums/http-status.enum';

import { FspAttributes } from '@121-service/src/fsps/enums/fsp-attributes.enum';
import {
  FspConfigurationProperties,
  Fsps,
} from '@121-service/src/fsps/enums/fsp-name.enum';
import { TransactionStatusEnum } from '@121-service/src/payments/transactions/enums/transaction-status.enum';
import { SeedScript } from '@121-service/src/scripts/enum/seed-script.enum';
import {
  createAndStartPayment,
  getTransactions,
  importFspReconciliationData,
  waitForPaymentTransactionsToComplete,
} from '@121-service/test/helpers/program.helper';
import { postProgramFspConfiguration } from '@121-service/test/helpers/program-fsp-configuration.helper';
import { seedPaidRegistrations } from '@121-service/test/helpers/registration.helper';
import {
  getAccessToken,
  resetDB,
} from '@121-service/test/helpers/utility.helper';
import {
  programIdWesteros,
  registrationWesteros1,
  registrationWesteros2,
  registrationWesteros3,
} from '@121-service/test/registrations/pagination/pagination-data';

describe('Reconciliate excel FSP data', () => {
  let accessToken: string;
  // Payment info
  const transferValue = 10;
  const paymentId = 1;
  const differentMatchColumn = 'accountId';

  // Registrations
  const registrationsWesteros = [
    registrationWesteros1,
    registrationWesteros2,
    registrationWesteros3,
  ];
  const referenceIdsWesteros = registrationsWesteros.map(
    (registration) => registration.referenceId,
  );
  const phoneNumbersWesteros = registrationsWesteros.map(
    (registration) => registration.phoneNumber,
  );

  const waitingTransactionStatusses = Array(registrationsWesteros.length).fill(
    TransactionStatusEnum.waiting,
  );

  const matchColumn = FspAttributes.phoneNumber;
  const statusColumn = 'status';

  const getTransactionStatusses = async () => {
    // Validate that transactions are still waiting
    const transactionsResponse = await getTransactions({
      programId: programIdWesteros,
      paymentId,
      accessToken,
    });
    return transactionsResponse.body.map((transaction) => transaction.status);
  };

  // No need to reset DB before each test, as we will only import reconciliation files unsuccessfully

  beforeAll(async () => {
    await resetDB(SeedScript.testMultiple, __filename);
    accessToken = await getAccessToken();

    await seedPaidRegistrations(
      registrationsWesteros,
      programIdWesteros,
      transferValue,
      [TransactionStatusEnum.waiting],
    );

    const excelFspConfigWithDifferentMatchColumn = {
      fspName: Fsps.excel,
      name: 'different-match-column-fsp',
      label: {
        en: 'Excel FSP config with different match column',
      },
      properties: [
        {
          name: FspConfigurationProperties.columnsToExport,
          value: [differentMatchColumn],
        },
        {
          name: FspConfigurationProperties.columnToMatch,
          value: differentMatchColumn,
        },
      ],
    };
    await postProgramFspConfiguration({
      programId: programIdWesteros,
      body: excelFspConfigWithDifferentMatchColumn,
      accessToken,
    });
  });

  it('Should throw an error when a payment is in progress', async () => {
    // Arrange
    const reconciliationData = [
      {
        [matchColumn]: registrationWesteros1.phoneNumber,
        [statusColumn]: TransactionStatusEnum.success,
      },
    ];

    const secondPayment = await createAndStartPayment({
      programId: programIdWesteros,
      transferValue,
      accessToken,
      referenceIds: referenceIdsWesteros,
    });
    // Act
    const importResult = await importFspReconciliationData(
      programIdWesteros,
      paymentId,
      accessToken,
      reconciliationData,
    );

    // Wait for payment transactions to complete, so it does not interfere with other tests
    await waitForPaymentTransactionsToComplete({
      programId: programIdWesteros,
      paymentId: secondPayment.body.id,
      accessToken,
      maxWaitTimeMs: 3_000,
      completeStatusses: [TransactionStatusEnum.waiting],
      paymentReferenceIds: referenceIdsWesteros,
    });

    const transactionStatuses = await getTransactionStatusses();

    // Assert
    expect(importResult.statusCode).toBe(HttpStatus.BAD_REQUEST);
    expect(importResult.body).toMatchSnapshot();
    // Expect that all transactions are still waiting after failed reconciliation attempts - no changes should be made
    expect(transactionStatuses).toEqual(waitingTransactionStatusses);
  });

  it('Should give an error when status column is missing', async () => {
    // Arrange
    const reconciliationData = [
      {
        [matchColumn]: registrationWesteros1.phoneNumber,
      },
    ];

    // Act
    const importResult = await importFspReconciliationData(
      programIdWesteros,
      paymentId,
      accessToken,
      reconciliationData,
    );
    const transactionStatuses = await getTransactionStatusses();

    // Assert
    expect(importResult.statusCode).toBe(HttpStatus.BAD_REQUEST);
    expect(importResult.body).toMatchSnapshot();
    // Expect that all transactions are still waiting after failed reconciliation attempts - no changes should be made
    expect(transactionStatuses).toEqual(waitingTransactionStatusses);
  });

  it('Should give an error when status column has invalid values', async () => {
    // Arrange
    const reconciliationData = [
      {
        [matchColumn]: registrationWesteros1.phoneNumber,
        [statusColumn]: 'invalid-status',
      },
    ];

    // Act
    const importResult = await importFspReconciliationData(
      programIdWesteros,
      paymentId,
      accessToken,
      reconciliationData,
    );
    const transactionStatuses = await getTransactionStatusses();

    // Assert
    expect(importResult.statusCode).toBe(HttpStatus.BAD_REQUEST);
    expect(importResult.body).toMatchSnapshot();
    // Expect that all transactions are still waiting after failed reconciliation attempts - no changes should be made
    expect(transactionStatuses).toEqual(waitingTransactionStatusses);
  });

  it('Should give an error when importing over 10.000 records', async () => {
    // Arrange
    const reconciliationData: Record<string, any>[] = [];
    for (let i = 0; i < 10001; i++) {
      reconciliationData.push({
        [matchColumn]: registrationWesteros1.phoneNumber,
        [statusColumn]: TransactionStatusEnum.success,
      });
    }

    // Act
    const importResult = await importFspReconciliationData(
      programIdWesteros,
      paymentId,
      accessToken,
      reconciliationData,
    );
    const transactionStatuses = await getTransactionStatusses();

    // Assert
    expect(importResult.statusCode).toBe(HttpStatus.BAD_REQUEST);
    expect(importResult.body).toMatchSnapshot();
    // Expect that all transactions are still waiting after failed reconciliation attempts - no changes should be made
    expect(transactionStatuses).toEqual(waitingTransactionStatusses);
  });

  it('Should give an error when match there is more than one match column in the import', async () => {
    // Arrange
    const matchColumn1 = FspAttributes.phoneNumber;
    const matchColumn2 = differentMatchColumn;

    const reconciliationData = [
      {
        [matchColumn1]: registrationWesteros1.phoneNumber,
        [statusColumn]: TransactionStatusEnum.success,
      },
      {
        [matchColumn2]: 'something random that does not matter',
        [statusColumn]: TransactionStatusEnum.success,
      },
    ];

    // Act
    const importResult = await importFspReconciliationData(
      programIdWesteros,
      paymentId,
      accessToken,
      reconciliationData,
    );
    const transactionStatuses = await getTransactionStatusses();

    // Assert
    expect(importResult.statusCode).toBe(HttpStatus.BAD_REQUEST);
    expect(importResult.body).toMatchSnapshot();
    // Expect that all transactions are still waiting after failed reconciliation attempts - no changes should be made
    expect(transactionStatuses).toEqual(waitingTransactionStatusses);
  });

  it('Should throw an error when there are duplicate values in the match column', async () => {
    // Arrange

    const reconciliationData = [
      {
        [matchColumn]: registrationWesteros1.phoneNumber,
        [statusColumn]: TransactionStatusEnum.success,
      },
      {
        [matchColumn]: registrationWesteros1.phoneNumber, // duplicate
        [statusColumn]: TransactionStatusEnum.success,
      },
    ];

    // Act
    const importResult = await importFspReconciliationData(
      programIdWesteros,
      paymentId,
      accessToken,
      reconciliationData,
    );
    const transactionStatuses = await getTransactionStatusses();

    // Assert
    expect(importResult.statusCode).toBe(HttpStatus.BAD_REQUEST);
    expect(importResult.body).toMatchSnapshot();
    // Expect that all transactions are still waiting after failed reconciliation attempts - no changes should be made
    expect(transactionStatuses).toEqual(waitingTransactionStatusses);
  });

  it('Should throw an error when importing data that is related to registrations that do not have the same Fsp config', async () => {
    // Arrange
    const reconciliationData = [
      {
        [matchColumn]: phoneNumbersWesteros[0],
        [statusColumn]: TransactionStatusEnum.success,
      },
      {
        [matchColumn]: phoneNumbersWesteros[2], // the registrations has a different FSP
        [statusColumn]: TransactionStatusEnum.success,
      },
    ];
    // Act
    const importResult = await importFspReconciliationData(
      programIdWesteros,
      paymentId,
      accessToken,
      reconciliationData,
    );
    const transactionStatuses = await getTransactionStatusses();

    // Assert
    expect(importResult.statusCode).toBe(HttpStatus.BAD_REQUEST);
    expect(importResult.body).toMatchSnapshot();
    // Expect that all transactions are still waiting after failed reconciliation attempts - no changes should be made
    expect(transactionStatuses).toEqual(waitingTransactionStatusses);
  });

  it('Should throw an error when no transactions are found for the the import', async () => {
    // Arrange

    const reconciliationData = [
      {
        [matchColumn]: '0000000000',
        [statusColumn]: TransactionStatusEnum.success,
      },
    ];

    // Act
    const importResult = await importFspReconciliationData(
      programIdWesteros,
      paymentId,
      accessToken,
      reconciliationData,
    );
    const transactionStatuses = await getTransactionStatusses();

    // Assert
    expect(importResult.statusCode).toBe(HttpStatus.BAD_REQUEST);
    expect(importResult.body).toMatchSnapshot();
    // Expect that all transactions are still waiting after failed reconciliation attempts - no changes should be made
    expect(transactionStatuses).toEqual(waitingTransactionStatusses);
  });
});
