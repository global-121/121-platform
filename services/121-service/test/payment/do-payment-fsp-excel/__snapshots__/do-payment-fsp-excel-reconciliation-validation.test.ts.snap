// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Reconciliate excel FSP data Should give an error when importing over 10.000 records 1`] = `
[
  "Too many records. Maximum number of records is 10000. You have 10001 records.",
]
`;

exports[`Reconciliate excel FSP data Should give an error when match there is more than one match column in the import 1`] = `
{
  "errors": "Multiple match columns found in the import file. Please only use one of the following match columns: phoneNumber, accountId.",
}
`;

exports[`Reconciliate excel FSP data Should give an error when status column has invalid values 1`] = `
{
  "message": "Invalid status value 'invalid-status'. Allowed values are: success, error",
  "statusCode": 400,
}
`;

exports[`Reconciliate excel FSP data Should give an error when status column is missing 1`] = `
{
  "message": "The status column is missing. Please make sure the column is named 'status'",
  "statusCode": 400,
}
`;

exports[`Reconciliate excel FSP data Should throw an error when a payment is in progress 1`] = `
{
  "message": "Cannot import FSP reconciliation data while payment is in progress",
  "statusCode": 400,
}
`;

exports[`Reconciliate excel FSP data Should throw an error when importing data that is related to registrations that do not have the same Fsp config 1`] = `
{
  "message": "Expected imported reconciliation to be related only to one FSP configuration, but found multiple",
  "statusCode": 400,
}
`;

exports[`Reconciliate excel FSP data Should throw an error when no transactions are found for the the import 1`] = `
{
  "message": "No matches found in relation to imported reconciliation",
  "statusCode": 400,
}
`;

exports[`Reconciliate excel FSP data Should throw an error when there are duplicate values in the match column 1`] = `
{
  "errors": "The match column 'phoneNumber' contains duplicate values.",
}
`;
